mozilla/teach.mozilla.org
# Badges

Fixes a ton of issues, although there are still a few things left to clean up. This is a port on top of the most recent develop as of 19 april 2016, with some "fixes" to gulp as well, to get rid of the static file generation, and a fix to route matching so that pages with :params resolve correctly.

Ping me for instructions on how to test this, because it requires having teach-api running on the badges branch, too, which requires some mozilla Credly credentials.

### I had to remove spider testing

This build system is still *very* weird and the spider task in combination with the static server (which doesn't even make sense anymore now that we don't static generate) would crash and burn trying to evaluate, because the static file build tasks got removed, because they *can't run* for urls with parameters in them.

I'm not happy with that, but I don't see a way out here - a lot of this code seems to have been set up in the assumption that we'd never have dynamic routes, which is now hitting us hard.

Followup to reinstroduce this filed as https://github.com/mozilla/teach.mozilla.org/issues/1836

### How to test this PR

As this PR relies on the Teach-API, you will need a copy of that running, checked out to the "badges" branch. This also requires having a working login system set up, so here's how I have it running:

1. psql running in the background
2. login.webmaker.org running locally
3. id.webmaker.org running locally
4. teach-api running on the `badges` branch
5. teach.mozilla.org running on the `badges` branch

However, there is now also an heroku app instance of the teach api with the badges branch, so you can probably tell your local teach.mozilla.org to use [https://teach-api-with-credly-badges.herokuapp.com](https://teach-api-with-credly-badges.herokuapp.com) as the TEACH_API endpoint, and things *should* work, with logins based on staging id.wmo

### Things to check

- Does the design look like the comps over on https://drive.google.com/file/d/0B2irLVDJivhLOEIzWkZiM3c5ZVE/view

#### flow when not logged in
1. don't be logged in
2. navigate to the "credentials" section of the site
  - [x] do badges load?
  - [x] can you navigate to a badge page?
  - [x] are you prompted to sign in?
  - [x] does the sign in button work?
    - [x] lets you sign in
    - [x] redirects to the page you were on
3. on individual badge pages
  - [x] does the badge load?
  - [x] are you prompted to sign in?
  - [x] does the sign in button work?
    - [x] lets you sign in
    - [x] redirects to the page you were on
  - [x] there are prev/next navigation entries at the bottom of the page
    - [x] they actually load what they claim they do

#### flow when logged in, but not linked to credly

1. be logged in
2. navigate to the "credentials" section of the site
  - [x] do badges load?
  - [x] can you navigate to a badge page?
3. on individual badge pages
  - [x] does the badge load?
  - [x] are you prompted to link with credly?
  - [x] does the link button work?
    - [x] lets you fill in an email and password (I recommend testing with a filtering email address, so something like `mavis+test1@mofo` so we can easily wipe all test accounts when we're happy that it works)
    - [x] returns you to the page you were on, and reloads it
    - [x] reload shows you as logged in, and credly-linked

#### flow when logged in, and linked to credly

2. navigate to the "credentials" section of the site
  - [x] do badges load?
  - [ ] do badges indicate their earned status? ("elligible", "pending", "achieved")
3. on individual badge pages
  - [x] does the badge load?
  - [ ] does it indicate the earned status? ("elligible", "pending", "achieved")
  - [ ] do you get an application form?
    - [x] text area
      - [x] typing text works
    - [x] URL text field
      - [x] typing a URL works
    - [ ] file attachment button
      - [ ] attaching a file works
  - [x] does the apply button work?
    - [x] disabled until you fill in something for evidence
    - [x] modal when you apply
    - [ ] after applying, the badge page reloads and the earned status is now "pending".

# issues addressed

fixes #1832
fixes #1823
fixes #1809
fixes #1777
fixes #1773
fixes #1770
fixes #1769
fixes #1722
fixes #1721
fixes #1720
fixes #1719
fixes #1718
fixes #1717
fixes #1714 - follow up to ensure file attachments work properly
fixes #1713
fixes #1712 - disabled social icons as per https://github.com/mozilla/teach.mozilla.org/pull/1830#issuecomment-216989101, follow up to reimplement
fixes #1711
fixes #1709
fixes #1704
fixes #1337
fixes #1192
